test/ex_doc/cli_test.exs

```diff
@@ -35,6 +35,41 @@ defmodule ExDoc.CLITest do
     assert io == "ExDoc v#{ExDoc.version()}\n"
   end
 
+  describe "--warnings-as-errors" do
+    test "exits with code 0 when no warnings" do
+      {[html, epub], _io} = run(["ExDoc", "1.2.3", @ebin, "--warnings-as-errors"])
+
+      assert html ==
+               {"ExDoc", "1.2.3",
+                [formatter: "html", apps: [:ex_doc], source_beam: @ebin, warnings_as_errors: true]}
+
+      assert epub ==
+               {"ExDoc", "1.2.3",
+                [formatter: "epub", apps: [:ex_doc], source_beam: @ebin, warnings_as_errors: true]}
+    end
+
+    test "exits with 1 with warnings" do
+      ExDoc.WarningCounter.increment()
```

Comment:
This test is running in _async_ mode, and you're modifying a global state here that could affect other tests, which is what happening with `test/mix/tasks/docs_test.exs`

Reply:
What would be the simplest way to solve this?
I have tried some but it requires a large refactoring.

Reply:
If I do a `ExDoc.WarningCounter.reset()` on the tests that use this global state would that work?

Reply:
> If I do a `ExDoc.WarningCounter.reset()` on the tests that use this global state would that work?

the issue is that since they are in separate files they could be executed at the same time.
If you execute reset in one file, it will be resetting the ones in the other tests. The only guarantee is that if you use `async: false` is that the tests within the same file are not going to be running concurrently.

Reply:
Sounds like this is going to be more work than I was expecting to get added 😞  I started a new job and am no longer working with elixir day to day so I likely won't be finishing this up anymore.  If someone else wants to pick it up feel free.

Reply:
> Sounds like this is going to be more work than I was expecting to get added disappointed I started a new job and am no longer working with elixir day to day so I likely won't be finishing this up anymore. If someone else wants to pick it up feel free.

Sure. No worries. I was already working with your PR so I'm familiar with it.
Thanks a lot for the PR, it is almost there.